### ISSUE TITLE
Controlling iiif min/max levels or extending

### ISSUE THREAD

**Issue by oblak1** [openseadragon/openseadragon#2363]
I am using OSD together with iiif image server. There are multiple images loaded together in the viewer/world, which means when the user zooms out to see them all, the images disappear for a second - until the server returns new smaller images - and I would like to avoid that by either loading a thumbnail there or limiting sizes/levels _since there is no point of loading eg. full 113x113px image if I already have a full 200x200px._
Here are a few questions regarding this:
 
1. Is it possible to have something like minLevel settings (similar to custom tile loading), where I could ignore sizes smaller than some number? Or, for example, load thumbnails from another url when the tiles are very small and there is no point in loading more of them (when they equal  1 tile is whole image).
2. By setting only one scaleFactors property (eg: 4) it still uses multiple levels. Can I maybe control number of levels with properties in info.json or is there another way of doing this?
3. Does OSD allow to extend its code (without forking the whole repo) and write my own [iiiftilesource.js](https://github.com/openseadragon/openseadragon/blob/master/src/iiiftilesource.js) incase there aren't any other solutions? If yes, can you quickly summarise how I would do this?

Thank you

**Comment by iangilman** [comment]
I know this isn't exactly what you're asking about, but I feel like by default, OSD won't make the image disappear when you zoom out... It should continue to show the lower res version. What settings are you using? I suppose `immediateRender` might cause this?

**Comment by iangilman** [comment]
Thank you for sharing the demo. Sure enough, this does appear to be vanilla OSD behavior... I would definitely consider this a bug! 

I've poked around at it a bit, and I'm actually not quite sure what causes it. Surely it's somewhere in the `TiledImage` code. I thought it might have to do with coverage, or the loading flag, or the levels interval, and maybe it does, but I couldn't chase it down. Basically, at some point, it decides that it needs the smaller tile and that it's not gonna bother drawing the bigger one. 

You can work around it, if you want, by just deleting any of the entries in the `sizes` array that you don't want it to load.

**Comment by oblak1** [comment]
Okay, thanks for looking at it. So I tried leaving only one sizes entry but it doesn't actually change the loading process (tried it on the demo above). I'm guessing it has more to do with what zoom level is being set. Or should I also delete something else?

Regarding the bug, If I find something I'll let you know. Although it seems strange to me as to why the zoom-in works as intended and this happens only on zoom-out, I'm guessing this is considered as the same event, meaning it should be handled the same?

**Comment by iangilman** [comment]
> Okay, thanks for looking at it. So I tried leaving only one sizes entry but it doesn't actually change the loading process (tried it on the demo above). I'm guessing it has more to do with what zoom level is being set. Or should I also delete something else?

Sure enough! I guess I didn't test it with throttling. I've now confirmed that you can fix it by adding `minLevel: 3,` to your `image` object. This tells OSD that there is nothing further toward the tip of the pyramid. 

> Regarding the bug, If I find something I'll let you know. Although it seems strange to me as to why the zoom-in works as intended and this happens only on zoom-out, I'm guessing this is considered as the same event, meaning it should be handled the same?

I don't think it has anything to do with zooming in or zooming out, per se... The TiledImage drawer just draws wherever it happens to be. In this case, it's small enough in the viewer that it's expecting a particular tile level, and for some reason, it doesn't bother drawing the tile levels below that when it's missing the tile that it wants. 

In the zoomed in case, if it doesn't have the high res tile it wants, it will still draw the low res tile. So there's an asymmetry there. 

At some point, we actually used to load the entire tip of the pyramid, but at some point we changed to loading the biggest tile that has the entire image in it. Perhaps there's some code that still assumes that we always load the entire tip.

**Comment by pearcetm** [comment]
In `TiledImage` there's a calculation of which levels to consider for drawing that does not depend on whether anything has been loaded already or not:

https://github.com/openseadragon/openseadragon/blob/ebab356c207e626b6622f88ffcb0cd28b918f85d/src/tiledimage.js#L1152-L1175

My guess this is the root of the problem.

**Comment by pearcetm** [comment]
> In the zoomed in case, if it doesn't have the high res tile it wants, it will still draw the low res tile. So there's an asymmetry there.

Based on the `_getLevelsInterval` code, the asymmetry here is because `highestLevel` depends on the current zoom level (from `this._scaleSpring.current.value` via `currentZeroRatio`) but `lowestLevel` does not depend on zoom. The logic is set up to be restrictive of how high-res a level should be requested (which makes sense as you don't want to be requesting very high res images when zoomed way out), but when zooming way out this is backfiring because already-loaded tiles that can cover the entire image are ignored.

**Comment by oblak1** [comment]
> Sure enough! I guess I didn't test it with throttling. I've now confirmed that you can fix it by adding minLevel: 3, to your image object. This tells OSD that there is nothing further toward the tip of the pyramid.

Nice one! This setting works for limiting the levels / sizes.

Regarding the zooming and levels I have a sub question regarding performance. In the above demo, by having set a `defaultZoomLevel`  it loads the necessary tiles for this level but it also loads the full image which is used at level 3 in this case (which you can see in the Network tab - the one with the ._../full/886,/default.jpg_ link I think). Is this an intentional behaviour? I'm guessing OSD loads the image as if the zoom level wasn't set and after that it loads the images for the current level - funnily enough it does help with the blinking since it doesn't have be loaded when zooming out.

> Based on the _getLevelsInterval code, the asymmetry here is because highestLevel depends on the current zoom level (from this._scaleSpring.current.value via currentZeroRatio) but lowestLevel does not depend on zoom. The logic is set up to be restrictive of how high-res a level should be requested (which makes sense as you don't want to be requesting very high res images when zoomed way out), but when zooming way out this is backfiring because already-loaded tiles that can cover the entire image are ignored.

Nice catch! I'm still trying to wrap my head around it but looking through the code and your comment if I write it in my words: it sets the `highestLevel` to the current level which tells the OSD to only look inside these bounds for tiles (eg. min-2 level instead of min-max level when requesting level 2). This would be an even more apparent issue if the default level 3 in my comment above wouldn't load right away.

Although I tried hardcoding the `highestLevel` to be always 5 and I don't get the desired results so maybe I'm missing something.

**Comment by pearcetm** [comment]
In `_updateViewport` there's another place that `minPixelRatio` is used to not draw levels that are "too high res":

https://github.com/openseadragon/openseadragon/blob/ebab356c207e626b6622f88ffcb0cd28b918f85d/src/tiledimage.js#L1225-L1231

I guess this is why hardcoding `highestLevel = 5` didn't fix the issue. You can try setting `minPixelRatio: 0.01` to see if that helps (you wouldn't need to hardcode `highestLevel` in that case because of how the calculation is done). Note that for real usage you wouldn't want to set `minPixelRatio` this low because it would load super high res tiles and have bad performance. This is just for testing.

**Comment by oblak1** [comment]
> I guess this is why hardcoding highestLevel = 5 didn't fix the issue. You can try setting minPixelRatio: 0.01 to see if that helps (you wouldn't need to hardcode highestLevel in that case because of how the calculation is done). Note that for real usage you wouldn't want to set minPixelRatio this low because it would load super high res tiles and have bad performance. This is just for testing.

I've tried setting this up but it doesn't help much because it is just using the levels for longer when zooming out and not using them while loading smaller ones.

I'm guessing the logic for updating viewport should be changed in the way to know what direction we are moving and have for loop change direction there (and also limit the lowerLevel then?) to have a similar logic as for selecting highres tiles. The for loop change (meaning going ++ instead of --) is probably necessary so the lower level tiles get priority.

https://github.com/openseadragon/openseadragon/blob/ebab356c207e626b6622f88ffcb0cd28b918f85d/src/tiledimage.js#L1216-L1216

Although this is just speculation since I haven't really managed to make anything work in this time. I would be happy to help with this but I think this is a bit of a difficult PR for me, changing the core updateViewport logic. 

Do you think this is the right direction? Would you have time to maybe create a PR?

**Comment by pearcetm** [comment]
> Do you think this is the right direction?

It seems pretty complicated - but then again it is a complicated issue.

My feeling is that, instead of the approach you describe, it would make sense to calculate the "best" level for a given zoom (where "best" does NOT necessarily equal highest resolution) and give that priority for loading and drawing, while allowing the viewer to make use of "too high-res" tiles that already exist (until the lower res "best" image is downloaded).

I'll see if I can play around with it when I resume working on https://github.com/openseadragon/openseadragon/pull/2310 which includes refactoring the drawing code already.

**Comment by iangilman** [comment]
One thing I should point out is we have some terminology confusion in the API. Tile pyramids have levels (like level 3 of the IIIF image @oblak1 is using) which are distinct sets of pixels. We also have options like `defaultZoomLevel` which just refer to the zoom value and have nothing to do with the pyramid levels. I suppose at some point we should consider changing those option names to be less confusing. Anyway, if you set the `defaultZoomLevel` to 3, you are zooming in 3 times from the default, not zooming into wherever the pyramid level 3 would be shown. 

In general, OSD is designed to hide the workings of the pyramid from the API... In most cases, you shouldn't have to think about the pyramid's details. Of course, for this issue, we're looking under the hood. 

@pearcetm I thought it would have to do with `_getLevelsInterval` so I tried hardcoding values in it, and it didn't seem to make any difference. I also tried disabling the `minPixelRatio` check, but now that I think of it, I may not have done them both at the same time. We're probably on the right track, but we haven't uncovered the whole story yet. 

> My feeling is that, instead of the approach you describe, it would make sense to calculate the "best" level for a given zoom (where "best" does NOT necessarily equal highest resolution) and give that priority for loading and drawing, while allowing the viewer to make use of "too high-res" tiles that already exist (until the lower res "best" image is downloaded).

Yes, that seems exactly right to me. 

> I'll see if I can play around with it when I resume working on https://github.com/openseadragon/openseadragon/pull/2310 which includes refactoring the drawing code already.

That would be awesome!

**Comment by iangilman** [comment]
> Regarding the zooming and levels I have a sub question regarding performance. In the above demo, by having set a defaultZoomLevel it loads the necessary tiles for this level but it also loads the full image which is used at level 3 in this case (which you can see in the Network tab - the one with the .../full/886,/default.jpg link I think). Is this an intentional behaviour? I'm guessing OSD loads the image as if the zoom level wasn't set and after that it loads the images for the current level - funnily enough it does help with the blinking since it doesn't have be loaded when zooming out.

We always start with the loading the largest level that contains the entire image in a single tile (in the case of your image, that's level 3). We do that to create the illusion that the image is always a complete entity, and not composed of tiles. It also means that the user is seeing a version of the entire image faster, since it's quicker to load that one initial tile than it is to load all of the tiles for level 5.

We then load the tiles for level 4 and fade them in, and then the tiles for level 5. By going through the levels like that, the experience of "sharpening" feels smooth and continuous, again, maintaining the illusion that you're looking at a singular entity, rather than a series of tiles. 

You can disable that functionality by using the `immediateRender` option; then it will load level 5 without first loading levels 3 and 4. In that case, you'll see the tiles coming in, which draws attention to the fragmented nature of the image data.

**Comment by pearcetm** [comment]
A fix for this bug/undesirable behavior is included in https://github.com/openseadragon/openseadragon/pull/2310

**Comment by iangilman** [comment]
Awesome! :)

**Comment by oblak1** [comment]
Great! :D

**Comment by oblak1** [comment]
There seems to be an issue with the build from `rotation-seams` @pearcetm - seeing there is a PR to master. I don't know if this is bug appeared because of the fix for this issue above or not but basically there is some kind of glitchy beahviour where tiles aren't positioned right while the other tiles are still loading if the user is dragging the canvas. The version that I used is a build from the branch mentioned above (the issue with blinking is fixed otherwise). Here is the codepen https://codepen.io/Matic-Oblak/pen/QWJwBWp and here is the video of the issue:

You can replicate the issue if you throttle it to slow 3G, restart the pen and drag it around.

https://github.com/openseadragon/openseadragon/assets/23618471/5634286b-6211-49bf-bab9-11f1af805596

**Comment by pearcetm** [comment]
Interesting. Will look into it, but would love if you could help track it down too. From the video it looks like lower-level tile positioning isn't being updated even though it clearly is still being drawn...

**Comment by iangilman** [comment]
@oblak1 Please be aware that #2310 is mostly a giant refactor for our drawing code to support WebGL. The fact that @pearcetm has included a fix for this issue is just a side note. We're not expecting #2310 to be ready to merge for a number of months. 

If you're interested in an isolated fix for this issue now, perhaps @pearcetm can guide you in making a PR for just this issue. Otherwise, we'll have to wait until #2310 lands. 

Of course, if you're interested in the drawing code refactor and are up for helping @pearcetm refine it, that's great! Just making sure you have the full context.

**Comment by pearcetm** [comment]
> @oblak1 Please be aware that #2310 is mostly a giant refactor for our drawing code to support WebGL. The fact that @pearcetm has included a fix for this issue is just a side note. We're not expecting #2310 to be ready to merge for a number of months.

Indeed! This is a big refactor and the drawing logic was (and continues to be) pretty convoluted. I've tried to disentangle things to some degree, but sometimes there is complexity there that isn't easy to understand intuitively. So, I'm not surprised something like this cropped up.

@oblak1 You could try commenting out the following line from that PR (starting at line 1390) to see if it helps:
````
            // Stop the loop if lower-res tiles would all be covered by
            // already drawn tiles
            if (this._providesCoverage(this.coverage, level)) {
                break;
            }
````

If it does it doesn't necessarily make it the correct thing to do to fix it, but it would be useful information to have.

**Comment by oblak1** [comment]
@iangilman @pearcetm okay, I understand this part of a bigger update still in the works. I would be glad if this issue could be somehow separated although looking through the code I see parts of the `tiledImage.js ` were changed to accommodate the webGL changes so it's a bit hard to separate logic, but if it can be done in the sense of merging and filtering out few commits I could take a look. 

Before that, getting back to the bug, @pearcetm I commented out your suggestion but it didn't help. I took some time and found out that the problem is the level 4 tile `_positionTile` isn't being called because it returns true for !drawTile at line 1737, that is because here:
```

for (let i = 0; i < levelList.length; i++) {
            let level = levelList[i];
            var drawLevel = false;

            //Avoid calculations for draw if we have already drawn this
            var currentRenderPixelRatio = this.viewport.deltaPixelsFromPointsNoRotate(
                this.source.getPixelRatio(level),
                true
            ).x * this._scaleSpring.current.value;
            if (i === levelList.length - 1 || 
                (!haveDrawn && currentRenderPixelRatio >= this.minPixelRatio) ) {
                drawLevel = true;
                haveDrawn = true;
            } else if (!haveDrawn) {
                continue;
            }
...
```

The drawLevel (which is later same as drawTile) isn't set to true for i === 4 anymore as soon as you zoom in. Due to changes  the tile isn't disappearing until level 5 fully loads but it doesn't get drawn (repositioned) because of i equals 5 in the if statement.

Also, another thing I noticed is that _positionTile is being called all the time by putting a console.log("") there. (_This is maybe intentional or something that I accidentally did while testing_)

**Comment by pearcetm** [comment]
@oblak1 I suspect this is related to a change I made in calling `_updateTile` which I was testing out. I noticed that in the previous drawing pipeline, it was called this way:

https://github.com/openseadragon/openseadragon/blob/ebab356c207e626b6622f88ffcb0cd28b918f85d/src/tiledimage.js#L1420-L1431

The order of the first two arguments didn't match the function signature:

https://github.com/openseadragon/openseadragon/blob/ebab356c207e626b6622f88ffcb0cd28b918f85d/src/tiledimage.js#L1454-L1455

So I tried swapping the arguments to match the signature... perhaps we found the consequence of this change? I reverted this back to the way it had been before. See if it makes a difference please!

**Comment by oblak1** [comment]
Yes, by swapping this back it seems to be working, although it seems interesting to me why this was done like this.

**Comment by pearcetm** [comment]
> Yes, by swapping this back it seems to be working, although it seems interesting to me why this was done like this.

Agreed. It almost looks like it was a mistake made a long time ago that just happened to make the drawing happen correctly. Or switching the order of the arguments was intentional but the rationale was not documented.
